compiler/rustc_middle/src/ty/sty.rs

```diff
@@ -1959,27 +1959,16 @@ impl<'tcx> TyS<'tcx> {
         }
     }
 
-    pub fn simd_type(&self, tcx: TyCtxt<'tcx>) -> Ty<'tcx> {
-        match self.kind() {
-            Adt(def, substs) => def.non_enum_variant().fields[0].ty(tcx, substs),
-            _ => bug!("`simd_type` called on invalid type"),
-        }
-    }
-
-    pub fn simd_size(&self, _tcx: TyCtxt<'tcx>) -> u64 {
-        // Parameter currently unused, but probably needed in the future to
-        // allow `#[repr(simd)] struct Simd<T, const N: usize>([T; N]);`.
-        match self.kind() {
-            Adt(def, _) => def.non_enum_variant().fields.len() as u64,
-            _ => bug!("`simd_size` called on invalid type"),
-        }
-    }
-
     pub fn simd_size_and_type(&self, tcx: TyCtxt<'tcx>) -> (u64, Ty<'tcx>) {
         match self.kind() {
             Adt(def, substs) => {
                 let variant = def.non_enum_variant();
-                (variant.fields.len() as u64, variant.fields[0].ty(tcx, substs))
+                let f0_ty = variant.fields[0].ty(tcx, substs);
+
+                match f0_ty.kind() {
+                    Array(f0_elem_ty, f0_len) => (f0_len.eval_usize(tcx, ParamEnv::empty()) as u64, f0_elem_ty),
```

Comment:
@oli-obk I'm pretty unfamiliar with the compiler. What we want to do is return the type and number of lanes of our SIMD type so that:

```rust
i32x4(i32, i32, i32, i32);
```

and

```rust
struct i32x4([i32; 4]);
```

will both return `(4, i32)`.

Is the way I'm evaluating the `N` in `[T; N]` here the right approach? I see the old code was looking at the ABI and ensuring it's `Vector` while it's at it. I don't know how to do that at this level.

Reply:
This will work until someone gets fancy and uses `struct i32x<N: usize>([i32; N]);`. I'm not sure exactly what kind of problems that will have, but I think we'll at least need to pass in a `ParamEnv` instead of using `empty`

Reply:
Reviewing all the changes here... this may not actually be a problem in practice, as this function is only used after monomorphization, so this will work fine I think. If we start doing some checks on the generic code, this will likely ICE.

Reply:
Should I drop a comment in there to that effect?

Reply:
Yes, please do, maybe even link to this PR so anyone encountering this site with an ICE has some related info